ui/src/features/Editor/components/OverlayUI/components/Toolbox/index.tsx

```diff
@@ -78,6 +79,10 @@ const Toolbox: React.FC<Props> = ({
     if (!isMainWorkflow) {
       return tool.id !== "reader" && tool.id !== "writer";
     }
+
+    if (isMainWorkflow && nodes?.some((node) => node.type === "reader")) {
```

Comment:
Passing nodes around and checking some EACH time isn't really desirable. Exactly like `isMainworkflow`, create a single callback that takes in the nodes and checks if it already has a Reader, and then pass the boolean instead.